### ISSUE TITLE
Warning: Resource IDs will be non-final in Android Gradle Plugin version 5.0, avoid using them as annotation attributes [NonConstantResourceId]

### ISSUE THREAD

**Issue by joshfriend** [airbnb/epoxy#1067]
After upgrading to AGP 4.1, I get a _lot_ of these warnings for the layout IDs passed to `@EpoxyModelClass`:

```
Warning: Resource IDs will be non-final in Android Gradle Plugin version 5.0, avoid using them as annotation attributes [NonConstantResourceId]
```

Just curious what this is going to mean for Epoxy's future and how I can begin preparing

**Comment by elihart** [comment]
Great question, I don't think we have an answer yet completely. Library modules already have non final ids, which we work around by using the Butterknife gradle plugin to generate an R2 class that has final id values. This is how Epoxy is able to work in multi module projects currently. this may work in 5.0, but I am not sure.

You can avoid using layout ids in annotations by using @ModelView on View classes instead of `@EpoxyModelClass`, which is one approach for now.

I imagine 5.0 is quite a ways off still. In the long run Jetpack Compose will replace Epoxy.

**Comment by hwwi** [comment]
> You can avoid using layout ids in annotations by using @ModelView on View classes instead of @EpoxyModelClass, which is one approach for now.

@ModelView.defaultLayout  has same warning.

**Comment by elihart** [comment]
yes, but you don't need to use a default layout, and it is more efficient not to.

**Comment by joshfriend** [comment]
> In the long run Jetpack Compose will replace Epoxy

That was kindof the answer I was expecting, though Compose 1.0 also seems a long ways off.

Thanks for the suggestions.

**Comment by mradzinski** [comment]
I stumbled across this issue while looking for some other stuff. Non-static ID's have been around for ages, they're just gonna be spread to Application modules. R2 will keep working in the same way it has been working before.

This doesn't mean R2 should be used. The proper way to account for dynamic ids is through view binding, though that would be a breaking change because view binding would stop being optional and instead become mandatory.

**Comment by arpitnnd** [comment]
Do we have an answer to this yet?

I've been using a lot of different layouts via `EpoxyModelWithHolder` approach.
Recently,  I switched to view binding from Butterknife & Kotlin synthetics.

I also started using Android Studio Arctic Fox which requires the new AGP 7.0 & the warnings now concern me.
Despite it now being in Beta, switching to Compose is not quite feasible considering the scale of the project.
The project is a single module application.

Looking for ways to continue using Epoxy with AGP 7.0.

**Comment by elihart** [comment]
I believe the R2 approach will continue to work to enable static id values, at least long enough for Compose to become stable and have time to migrate to it.

You can also use Epoxy without any annotations with resource values by using custom views with @ModelView

**Comment by dinhlamvn** [comment]
@joshfriend In this case, I think you can use `View.inflate(..)` in `init` function on your ModelView and your layout will be contained by `merge` tag.

**Comment by minhnhattonthat** [comment]
or we can just override ```getDefaultLayout()``` in the ```EpoxyModelWithHolder``` class, since that's what the annotation does anyway.

**Comment by elihart** [comment]
> or we can just override getDefaultLayout() in the EpoxyModelWithHolder class, since that's what the annotation does anyway.

For some usages yes, although there are other places in Epoxy that use resources in annotations that are not so easily replaceable
